File: example/rust/src/bin/2_zero_and_read.rs
use rebotarm_control_rt_rust_examples::common::{
    has_flag, parse_port, prompt, B601Arm, B601_JOINTS,
};
use std::env;
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().skip(1).collect();
    if has_flag(&args, "--help") || has_flag(&args, "-h") {
        println!("Usage: cargo run --bin 2_zero_and_read -- --port /dev/ttyACM0 [--skip-zero]");
        return Ok(());
    }

    let arm = B601Arm::open(&parse_port(&args))?;
    println!("connected: {}", arm.port);

    if !has_flag(&args, "--skip-zero") {
        println!("This will set the current pose as zero for all B601 motors.");
        println!("Type YES to continue.");
        if prompt("confirm> ")?.as_deref() == Some("YES") {
            arm.disable()?;
            for (joint, motor) in B601_JOINTS.iter().zip(&arm.motors) {
                motor.set_zero_position()?;
                println!("zero set: {}", joint.name);
            }
        } else {
            println!("zero skipped");
        }
    }

    println!("Press Enter to read state again, q to quit.");
    loop {
        arm.print_state();
        let Some(line) = prompt("> ")? else {
            break;
        };
        if matches!(line.as_str(), "q" | "quit" | "exit") {
            break;
        }
    }

    arm.close();
    Ok(())
}

File: example/cpp/src/2_zero_and_read.cpp
#include "example_common.hpp"

#include <exception>
#include <iostream>

int main(int argc, char** argv) {
  try {
    if (example::has_flag(argc, argv, "--help") || example::has_flag(argc, argv, "-h")) {
      std::cout << "Usage: ./2_zero_and_read --port /dev/ttyACM0 [--skip-zero]\n";
      return 0;
    }

    auto arm = example::B601Arm::open(example::parse_port(argc, argv));
    std::cout << "connected: " << arm.port << "\n";

    if (!example::has_flag(argc, argv, "--skip-zero")) {
      std::cout << "This will set the current pose as zero for all B601 motors.\n";
      std::cout << "Type YES to continue.\n";
      if (example::prompt("confirm> ").value_or("") == "YES") {
        arm.disable();
        for (int i = 0; i < static_cast<int>(arm.motors.size()); ++i) {
          arm.motors[i].set_zero_position();
          std::cout << "zero set: " << example::b601_joints()[i].name << "\n";
        }
      } else {
        std::cout << "zero skipped\n";
      }
    }

    std::cout << "Press Enter to read state again, q to quit.\n";
    while (true) {
      arm.print_state();
      const auto line = example::prompt("> ");
      if (!line || *line == "q" || *line == "quit" || *line == "exit") break;
    }

    arm.close();
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
}
